esm/dom/stream.js

```diff
@@ -0,0 +1,180 @@
+import {Writable} from 'stream';
+import {WritableStream} from 'htmlparser2/lib/WritableStream';
+
+import {ELEMENT_NODE, SVG_NAMESPACE} from '../shared/constants';
+import {CUSTOM_ELEMENTS, PREV, END, VALUE} from '../shared/symbols';
+import {keys} from '../shared/object';
+
+import {knownBoundaries, knownSiblings} from '../shared/utils';
+import {attributeChangedCallback, connectedCallback} from '../interface/custom-element-registry.js';
+
+import {HTMLDocument} from '../html/document.js';
+import {SVGDocument} from '../svg/document.js';
+import {XMLDocument} from '../xml/document.js';
+
+/**
+ * @typedef {import('../interface/node').Node} Node
+ * @typedef {import('../svg/element').SVGElement} SVGElement
+ * @typedef {{ "text/html": HTMLDocument, "image/svg+xml": SVGDocument, "text/xml": XMLDocument }} MimeToDoc
+ */
+
+/**
+ * @template {keyof MimeToDoc} MIME
+ * @extends {Writable}
+ */
+export class DOMStream extends Writable {
+  /**
+   * @param {MIME} mimeType
+   * @param {(name: string, attributes: Record<string, string>) => boolean} filter
+   */
+  constructor (mimeType, filter) {
+    super();
+    this.mimeType = mimeType;
+    if (mimeType === 'text/html') this.isHTML = true;
+    this.filter = filter;
+    /**
+     * @type {{
+     *   document: MimeToDoc[MIME]
+     *   node: MimeToDoc[MIME]|Node
+     *   ownerSVGElement: SVGElement|undefined
+     *   rootNode: Node
+     * }[]}
+     */
+    this.stack = [];
+    this.init = this.init.bind(this);
+    this.init();
+  }
+
+  newDocument () {
+    let document;
+    if (this.mimeType === 'text/html') {
+      document = new HTMLDocument();
+    } else if (this.mimeType === 'image/svg+xml') {
+      document = new SVGDocument();
+    } else {
+      document = new XMLDocument();
+    }
+    if (this.doctype) document.doctype = this.doctype;
+    this.stack.push({ document, node: document });
+  }
+
+  init ()  {
+    this.parserStream = new WritableStream({
+      // <!DOCTYPE ...>
+      onprocessinginstruction: (name, data) => {
```

Comment:
could we share most of this code somehow instead of (looks like) repeating it all over? 🤔
I'm OK having a new file with repeated code for the time being, but I wonder why you didn't reuse methods already available elsewhere

Reply:
P.S. it is very possible that sharing code will also help with code coverage